src/sagemaker/telemetry/telemetry_logging.py

```diff
@@ -0,0 +1,229 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Telemetry module for SageMaker Python SDK to collect usage data and metrics."""
+from __future__ import absolute_import
+import logging
+import platform
+import sys
+from time import perf_counter
+from typing import List
+
+from sagemaker.utils import resolve_value_from_config
+from sagemaker.config.config_schema import TELEMETRY_OPT_OUT_PATH
+from sagemaker.telemetry.constants import (
+    Feature,
+    Status,
+    DEFAULT_AWS_REGION,
+)
+from sagemaker.user_agent import SDK_VERSION, process_studio_metadata_file
+
+logger = logging.getLogger(__name__)
+
+OS_NAME = platform.system() or "UnresolvedOS"
+OS_VERSION = platform.release() or "UnresolvedOSVersion"
+OS_NAME_VERSION = "{}/{}".format(OS_NAME, OS_VERSION)
+PYTHON_VERSION = "{}.{}.{}".format(
+    sys.version_info.major, sys.version_info.minor, sys.version_info.micro
+)
+
+TELEMETRY_OPT_OUT_MESSAGING = (
+    "SageMaker Python SDK will collect telemetry to help us better understand our user's needs, "
+    "diagnose issues, and deliver additional features.\n"
+    "To opt out of telemetry, please disable via TelemetryOptOut parameter in SDK defaults config. "
+    "For more information, refer to https://sagemaker.readthedocs.io/en/stable/overview.html"
+    "#configuring-and-using-defaults-with-the-sagemaker-python-sdk."
+)
+
+FEATURE_TO_CODE = {
+    str(Feature.SDK_DEFAULTS): 1,
+    str(Feature.LOCAL_MODE): 2,
+}
+
+STATUS_TO_CODE = {
+    str(Status.SUCCESS): 1,
+    str(Status.FAILURE): 0,
+}
+
+
+def _telemetry_emitter(feature: str, func_name: str):
+    """Decorator to emit telemetry logs for SageMaker Python SDK functions"""
+
+    def decorator(func):
+        def wrapper(self, *args, **kwargs):
```

Comment:
Why need to pass in the 'self' arg here?
This is not a class method which should not require a 'self' arg right?

I think we can retrieve the self from the first element of *args

Reply:
See the example: https://github.com/knikure/sagemaker-python-sdk/blob/c48d7c8c4d871dc4bc07ef773f0e58df6f1d4cc8/src/sagemaker/experiments/_utils.py#L87